File: packages/runtime-web-three/src/runtimePrefabsHierarchy.ts
import type { IBundleManifest, IPrefabsIr, ISystemsIr, IWorldIr } from "@threenative/ir";

import { loadSystemModule, runSchedule } from "./systems/runner.js";

export interface IRuntimePrefabsHierarchyReport {
  entities: Array<{ id: string; parent: string | null }>;
  schema: "threenative.runtime-prefabs-hierarchy";
  version: "0.1.0";
}

export async function traceRuntimePrefabsHierarchy(
  world: IWorldIr,
  systems: ISystemsIr,
  prefabs: IPrefabsIr,
  options: { bundlePath?: string; manifest?: IBundleManifest } = {},
): Promise<IRuntimePrefabsHierarchyReport> {
  const workingWorld = structuredClone(world);
  const module = options.bundlePath && options.manifest
    ? await loadSystemModule(options.bundlePath, options.manifest)
    : { systems: {} };

  await runSchedule({
    module,
    prefabs,
    schedule: "update",
    systems,
    world: workingWorld,
  });

  return {
    entities: workingWorld.entities
      .map((entity) => ({
        id: entity.id,
        parent: readParent(entity.components.Hierarchy),
      }))
      .sort((left, right) => left.id.localeCompare(right.id)),
    schema: "threenative.runtime-prefabs-hierarchy",
    version: "0.1.0",
  };
}

function readParent(value: unknown): string | null {
  if (typeof value === "object" && value !== null && !Array.isArray(value) && typeof (value as { parent?: unknown }).parent === "string") {
    return (value as { parent: string }).parent;
  }
  return null;
}

File: runtime-bevy/crates/threenative_runtime/src/runtime_prefabs_hierarchy.rs
use serde::Serialize;
use threenative_loader::LoadedBundle;

use crate::systems_context::NativeSystemTimeSnapshot;
use crate::systems_host::{NativeSystemsHostRun, run_native_systems_once};

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RuntimePrefabsHierarchyReport {
    pub entities: Vec<RuntimePrefabEntityReport>,
    pub schema: &'static str,
    pub version: &'static str,
}

#[derive(Debug, Serialize)]
pub struct RuntimePrefabEntityReport {
    pub id: String,
    pub parent: Option<String>,
}

pub fn trace_runtime_prefabs_hierarchy(bundle: &mut LoadedBundle) -> RuntimePrefabsHierarchyReport {
    let _run: NativeSystemsHostRun = run_native_systems_once(
        bundle,
        NativeSystemTimeSnapshot {
            delta: 0.016,
            dt: 0.016,
            elapsed: 1.0,
            fixed_delta: 0.016,
            fixed_dt: 0.016,
            paused: false,
        },
    )
    .expect("runtime prefab hierarchy systems should run");

    let mut entities = bundle
        .world
        .entities
        .iter()
        .map(|entity| RuntimePrefabEntityReport {
            id: entity.id.clone(),
            parent: entity
                .components
                .hierarchy
                .as_ref()
                .and_then(|hierarchy| hierarchy.parent.clone()),
        })
        .collect::<Vec<_>>();
    entities.sort_by(|left, right| left.id.cmp(&right.id));

    RuntimePrefabsHierarchyReport {
        entities,
        schema: "threenative.runtime-prefabs-hierarchy",
        version: "0.1.0",
    }
}
